### PR TITLE
New caching strategy

### PR DESCRIPTION
**Pull request by charig**
Summary of changes:
* Extend Code objects so that they also store the #different `<ldVar/stVar, env>` they contain. This enables to create binding caches of dynamic size eventually. With static size, it enables to optimize based on the fact that there are no collisions whenever this number is < cache size.      
* Extend compiler so that it assigns a unique number to every `<ldVar/stVar, env>`.
* Extend pir2rir so that it assigns a unique number to every `<ldVar/stVar, env>`. This enables to avoid the cache clearing we are performing after every environment change. 
* Extend rir bytecode so every load/stote has its corresponding (load/store)Cache version. The cache versions contain one extra immediate: the unique number calculated. The standard versions do not use the cache at all. This PR does not implement it, but this allow us to avoid all the caching overhead for loads and stores for which PIR can prove are not reused, etc.

TODOs:
* Avoid memsetting when pir does not need it.
* Move from a static array to a dynamic container (R stack mapping) initialized with the precise size needed. 
* It seems like mem setting is one of the most significant sources of performance. This seems to be different when mapping in the R stack. By mapping on the R stack we can also reuse in some RIR trampoline calls (loops) thus avoiding extra memsettings.
* Analysis at the PIR level to use standard instructions (no caching) whenever we know beforehand the values will not be reused enough times.
* What happens with caches fpr promises? Can we do better in PIR?

### CODE REVIEW

File: rir/src/compiler/translations/pir_2_rir/pir_2_rir.cpp [hunk 1]
```diff
@@ -29,438 +29,6 @@ namespace pir {
 
 namespace {
 
-/*
```
Comment (o-) [inline rir/src/compiler/translations/pir_2_rir/pir_2_rir.cpp:32]: did you move this stuff unchanged or are there changes in the code you moved into the other file?

File: rir/src/compiler/translations/pir_2_rir/pir_2_rir.cpp [hunk 2]
```diff
@@ -1336,8 +934,8 @@ rir::Code* Pir2Rir::compileCode(Context& ctx, Code* code) {
     });
     cb.flush();
 
-    auto localsCnt = alloc.slots();
-    auto res = ctx.finalizeCode(localsCnt);
+    auto res =
```
Comment (o-) [inline rir/src/compiler/translations/pir_2_rir/pir_2_rir.cpp]: I feel like some documentation get's lost with this change. why are we passing slots() ? in the old version we knew: because its the locals count...

File: rir/src/interpreter/interp.cpp [hunk 3]
```diff
@@ -1520,11 +1535,13 @@ SEXP evalRirCode(Code* c, InterpreterInstance* ctx, SEXP env,
 #endif
 
     assert(c->info.magic == CODE_MAGIC);
+    bool existingLocals = localsBase;
 
     BindingCache bindingCache[BINDING_CACHE_SIZE];
-    memset(&bindingCache, 0, sizeof(bindingCache));
+    bool smallCache = c->bindingsCount <= BINDING_CACHE_SIZE;
+    if (env != symbol::delayedEnv || existingLocals)
```
Comment (o-) [inline rir/src/interpreter/interp.cpp]: I don't think existingLocals allows you to not initialize the cache

Comment (o-) [reply rir/src/interpreter/interp.cpp]: can you add a comment that the zeroing is delayed to the setEnv, it took me a bit to understand it

File: rir/src/interpreter/interp.cpp [hunk 4]
```diff
@@ -1644,7 +1660,7 @@ SEXP evalRirCode(Code* c, InterpreterInstance* ctx, SEXP env,
                     }
                 }
             }
-
+            memset(&bindingCache, 0, sizeof(bindingCache));
```
Comment (o-) [inline rir/src/interpreter/interp.cpp:1672]: would be nice if we only zero affected bindings

File: rir/src/interpreter/interp.cpp [hunk 5]
```diff
@@ -1232,43 +1232,55 @@ static SEXP seq_int(int n1, int n2) {
...
 
-    Immediate cidx = idx % BINDING_CACHE_SIZE;
-    if (bindingCache[cidx].idx == idx) {
+    Immediate cidx;
+    if (smallCache)
+        cidx = cacheIdx;
+    else
+        cidx = cacheIdx & CACHE_MASK;
+
+    if (bindingCache[cidx].idx == poolIdx) {
```
Comment (o-) [inline rir/src/interpreter/interp.cpp]: so we changed the cache to contain entries from different environments (to avoid clearing them on change env).

so this is not correct, you would need to check if the environment is correct too. you would need to add the env the cache is for to the bindingCache cell

File: rir/src/ir/insns.h [hunk 6]
```diff
@@ -54,17 +54,29 @@ DEF_INSTR(ldfun_, 1, 0, 1, 0)
  */
 DEF_INSTR(ldvar_, 1, 0, 1, 0)
 
+/**
+ * ldvar_:: like ldvar.
+ * Stores an additional immediate with a unique number for the cache bindings.
+ */
+DEF_INSTR(ldvar_cache_, 2, 0, 1, 0)
```
Comment (o-) [inline rir/src/ir/insns.h]: I would have prefered a lot, if we would not double the instructions. instead cache idx of 0 (or -1) could mean no caching.

Comment (charig) [reply rir/src/ir/insns.h]: Ok, but do you have a strong reason? Because unless we have some "size" like argument it is usually better to have more instructions with particular semantics than adding an if to every load/store

File: rir/src/interpreter/interp.cpp (hunk 4 above)
Comment (charig) [reply rir/src/interpreter/interp.cpp:1672]: Yes, but far from straightforward...

File: rir/src/interpreter/interp.cpp (hunk 5 above)
Comment (charig) [reply rir/src/interpreter/interp.cpp]: Why? 
In the cache slot we store the unique number determining a pair<symbol,env>.

Comment (o-) [reply rir/src/interpreter/interp.cpp]: so assume `<"a", e1>` has id 1 and `<"a", e2>` has id 9. they will both end up in slot 1, the check will pass, since both refer to the symbol "a" and you will read the wrong binding.

File: rir/src/ir/insns.h (hunk 6 above)
Comment (o-) [reply rir/src/ir/insns.h]: less code duplication (they all do almost the same. I mean in pir you chose to use the same instruction) and most will have a cache

File: rir/src/compiler/translations/pir_2_rir/allocators.h [hunk 7]
```diff
@@ -0,0 +1,506 @@
...
+
+struct nameAndEnvHash {
+  public:
+    template <typename T, typename U>
+    std::size_t operator()(const std::pair<T, U>& x) const {
+        return std::hash<T>()(x.first) ^ std::hash<U>()(x.second);
+    }
+};
+
+class CachePositionAllocator {
```
Comment (o-) [inline rir/src/compiler/translations/pir_2_rir/allocators.h:442]: I didn't see this, because I did not see that you added new stuff to the file.

Why do you need to preprocess this? you should be able to do this on the fly. if you already saw a pair, return the id, otherwise assign a new id.

File: rir/src/interpreter/interp.cpp (hunk 5 above)
Comment (charig) [reply rir/src/interpreter/interp.cpp]: No, because `bindingCache[1].idx = 1` or `bindingCache[1].idx = 9` to distinguish both cases.

File: rir/src/ir/insns.h (hunk 6 above)
Comment (charig) [reply rir/src/ir/insns.h]: Ok, but the distinction was on purpose. PIR is not "executable" -> I do not mind executing ifs at compile time why I may mind execute an if !=-1 for every load/store RIR BC.

Oth, (all) have a cache now. But I do not think that will necessarily be the like that in future versions. Analysis may realize for lot of cases that the cache is not needed (for instance load stores outside loops)

File: rir/src/compiler/translations/pir_2_rir/allocators.h (hunk 7 above)
Comment (charig) [reply rir/src/compiler/translations/pir_2_rir/allocators.h:442]: Yes, when I started I thought I may need sth fancier, it ended up being pretty simple. Anyway, I do not think one more visitor pass harms at this point and the lowering is more readable like this.

Comment (o-) [reply rir/src/compiler/translations/pir_2_rir/allocators.h:442]: why is it more readable? you don't even need to change the api. just change this class here, to not do a run before

File: rir/src/interpreter/interp.cpp (hunk 5 above)
Comment (o-) [reply rir/src/interpreter/interp.cpp]: https://github.com/reactorlabs/rir/pull/471#discussion_r280716465

Comment (o-) [reply rir/src/interpreter/interp.cpp]: duh, sorry. I did not notice that you changed the index to be the cache index instead of the symbol

Comment (o-) [reply rir/src/interpreter/interp.cpp]: but the idx field does not contain the cacheIdx, but actually the poolIdx, which is just the symbol, not?

Comment (o-) [reply rir/src/interpreter/interp.cpp]: also, this check is not needed on small caches. so it should be
```
if (smallCache || bindingCache[cidx].idx == cacheIdx)
```

and then store the cacheIdx into the field everywhere where you update it

Comment (charig) [reply rir/src/interpreter/interp.cpp]: Ah ok, yeah, the only issue was it should store the cacheIdx yes.
The check is still needed right now because we have to avoid loading of uninit variables

File: rir/src/interpreter/interp.cpp [hunk 8]
```diff
@@ -1520,11 +1534,20 @@ SEXP evalRirCode(Code* c, InterpreterInstance* ctx, SEXP env,
 #endif
 
     assert(c->info.magic == CODE_MAGIC);
+    bool existingLocals = localsBase;
 
     BindingCache bindingCache[BINDING_CACHE_SIZE];
-    memset(&bindingCache, 0, sizeof(bindingCache));
+    // Position zero is not counted for smallcache because we can not
+    // distinguish slot zero from uninit.
```
Comment (o-) [inline rir/src/interpreter/interp.cpp:1541]: ehm, use cacheLocation-1 as the index?

File: rir/src/compiler/translations/rir_2_pir/rir_2_pir.cpp [hunk 9]
```diff
@@ -200,20 +200,22 @@ bool Rir2Pir::compileBC(const BC& bc, Opcode* pos, Opcode* nextPos,
         break;
 
     case Opcode::ldvar_:
-    case Opcode::ldvar_for_update_:
+    case Opcode::ldvar_cache_:
+    case Opcode::ldvar_for_update_cache_:
```
Comment (o-) [inline rir/src/compiler/translations/rir_2_pir/rir_2_pir.cpp:204]: only saw this now, but in general cached (with d) would be a much better name for all of them. because cache is used as adverb and not verb: perform a lookup, a cached one.

File: rir/src/interpreter/interp.cpp [hunk 10]
```diff
@@ -1232,43 +1232,54 @@ static SEXP seq_int(int n1, int n2) {
...
 
-    Immediate cidx = idx % BINDING_CACHE_SIZE;
-    if (bindingCache[cidx].idx == idx) {
+    Immediate cidx;
+    if (smallCache)
+        cidx = cacheIdx;
+    else
+        cidx = cacheIdx & CACHE_MASK;
+
+    if (bindingCache[cidx].idx == cacheIdx)
```
Comment (o-) [inline rir/src/interpreter/interp.cpp:1258]: i would go for
```
bool hit = smallCache || bindingCache[cidx].idx == cacheIdx;
if (hit && bindingCache[cidx].loc)
  ...
```

this should be a bit faster for small cache and additionally solve the zero index issue

File: rir/src/interpreter/interp.cpp (hunk 8 above)
Comment (o-) [reply rir/src/interpreter/interp.cpp:1541]: (or see suggestion at https://github.com/reactorlabs/rir/pull/471#discussion_r281169517)

File: rir/src/interpreter/interp.cpp [hunk 11]
```diff
@@ -3316,7 +3414,11 @@ SEXP evalRirCode(Code* c, InterpreterInstance* ctx, SEXP env,
         }
 
         INSTRUCTION(guard_fun_) {
+#ifndef UNSOUND_OPTS
```
Comment (o-) [inline rir/src/interpreter/interp.cpp]: why is this here?

File: rir/src/compiler/translations/pir_2_rir/allocators.h [hunk 12]
```diff
@@ -0,0 +1,470 @@
...
+            if (*i == v)
+                return false;
+        for (size_t i = argNumber + 1; i < instr->nargs(); ++i)
+            if (instr->arg(i).val() == v)
+                return false;
+        return true;
+    }
+};
+
+struct nameAndEnvHash {
```
Comment (o-) [inline rir/src/compiler/translations/pir_2_rir/allocators.h]: this is very strange. on the one hand you call this nameAndEnvHash, suggesting that this is just for this name/env pair. but then the actual implementation is a template function, which is only necessary if you want to reuse this.

please just use `pairhash<SEXP,Value*>` from common.h

File: rir/src/compiler/translations/pir_2_rir/allocators.h [hunk 13]
```diff
@@ -0,0 +1,470 @@
...
+
+class CachePositionAllocator {
+  public:
+    typedef size_t SlotNumber;
+
+    size_t numberOfBindings() { return uniqueNumbers.size(); }
+    SlotNumber slotFor(SEXP varName, Value* environment) {
+        std::pair<SEXP, Value*> key =
+            std::pair<SEXP, Value*>(varName, environment);
+        return uniqueNumbers.emplace(key, uniqueNumbers.size() + 1)
```
Comment (o-) [inline rir/src/compiler/translations/pir_2_rir/allocators.h:449]: this always overrides and never returns the same value?

File: rir/src/compiler/translations/pir_2_rir/allocators.h [hunk 14]
```diff
@@ -0,0 +1,470 @@
...
+    }
+};
+
+class CachePositionAllocator {
+  public:
+    typedef size_t SlotNumber;
+
+    size_t numberOfBindings() { return uniqueNumbers.size(); }
+    SlotNumber slotFor(SEXP varName, Value* environment) {
+        std::pair<SEXP, Value*> key =
```
Comment (o-) [inline rir/src/compiler/translations/pir_2_rir/allocators.h]: `nameAndEnv key`

File: rir/src/compiler/translations/pir_2_rir/allocators.h [hunk 15]
```diff
@@ -0,0 +1,470 @@
...
+            std::pair<SEXP, Value*>(varName, environment);
+        return uniqueNumbers.emplace(key, uniqueNumbers.size() + 1)
+            .first->second;
+    }
+
+  private:
+    typedef std::pair<SEXP, Value*> nameAndEnv;
```
Comment (o-) [inline rir/src/compiler/translations/pir_2_rir/allocators.h]: by convention we write types with first letter capitalized. `NameAndEnv`

**Review by o- — APPROVED** [review]

File: rir/src/compiler/translations/pir_2_rir/allocators.h (hunk 13 above)
Comment (charig) [reply rir/src/compiler/translations/pir_2_rir/allocators.h:449]: emplace does not insert if the key is already defined and returns that key

File: rir/src/interpreter/interp.cpp (hunk 10 above)
Comment (charig) [reply rir/src/interpreter/interp.cpp:1258]: `cacheLocation - 1` implies substracting on every access, I'd prefer to just not use the zero index, its only 16 bytes lost. 

Testing for `loc!=null` enables to use the zero index too, but it implies to add one more check for the nonSmall version (cacheIdx + loc != null) while it adds nothing to the smallcache (previously testing for idx now for loc)

File: rir/src/compiler/translations/pir_2_rir/allocators.h (hunk 13 above)
Comment (o-) [reply rir/src/compiler/translations/pir_2_rir/allocators.h:449]: nevermind

File: rir/src/ir/Compiler.cpp [hunk 16]
```diff
@@ -458,497 +470,497 @@ bool compileSpecialCall(CompilerContext& ctx, SEXP ast, SEXP fun, SEXP args_, bo
...
+        return true;
+    }
+    Else(break)
+}
 
-        // 3) Special case [ and [[
-
-        if (lhsParts.size() != 2) {
-            return false;
+// Find all parts of the lhs
```
Comment (mhyee) [inline rir/src/ir/Compiler.cpp:487]: Whoa, what happened to the indentation here?

Comment (o-) [reply rir/src/ir/Compiler.cpp:487]: hah, beatiful. yeah, I had to disable whitespaces in the diff settings to be able to review this PR at all. that's why I did not see it. I think this is the final argument why we should never run clang-format on the whole repo. @charig , can you please revert those whitespace changes?
